### ISSUE TITLE
Map object model to document model.

### ISSUE THREAD

**Issue by kdbanman** [CMPUT301W15T07/TravelTracker#47]
This is mostly a future problem, but it has some relevance to our UML project 3 deliverable, since membership relationships for our data model classes .

We should figure out how each of our model objects (Claim, Tag, User, etc) should map to JSON documents.  A [relevant comment](https://github.com/CMPUT301W15T07/TravelTracker/issues/24#issuecomment-75187530) by @TheLastBanana pointed out that we should use ElasticSearch [parent-child](http://www.elasticsearch.org/guide/en/elasticsearch/guide/current/parent-child-mapping.html) relationships.  I think this is correct.  

We already have access to an ElasticSearch *index* (http://cmput301.softwareprocess.es:8080/cmput301w15t07/), so what's left for us is:

1. Decide which model objects are the actual [documents](http://www.elasticsearch.org/guide/en/elasticsearch/guide/current/mapping.html).
2. Decide how they are related.  (parents? children?)
3. Define their properties and config options.

-----------------------------------

my thoughts:

1) Say our documents types are User, Claim, Item, and Tag.  The rest of the data model objects would be stored as parts of those documents.  **Should we have a document type for each data model class?**

2) Claims are children of Users, and Items children of Claims.  We tell this to ElasticSearch with

```javascript
PUT /cmput301w15t07
{
  "mappings": {
    "user": { },
    "claim": {
      "_parent": {
        "type": "user" 
      },
    "item": {
      "_parent": {
        "type": "claim" 
      }
    }
  }
}
```

**But what about Tags?**  They have a many to ~~one~~ many relationship with User and with Claim.  That isn't a problem in Java - just put a `claimID` and a `userID` attribute in the Tag class.  But ElasticSearch enforces one parent per child.

3) Starter example for the claim document type is [here](http://www.codeshare.io/FqRqx), as would be used with `curl` and ElasticSearch's HTTP API.  This step should be a tedious fill in the blanks chore after the above 2 steps are addressed.

**Comment by TheLastBanana** [comment]
Tags are in a many-to-many relationship with Claims, so our convention doesn't really apply anyway. A Tag's parent is its User, so the Tag should have a userID. As for which Claims it applies to, either the Claim needs a list of Tags or vice-versa.

**Comment by kdbanman** [comment]
> Tags are in a many-to-many relationship with Claims

Woops, silly mistake.  Thank you.  

> either the Claim needs a list of Tags or vice-versa

That's clear now.  Do you see a benefit one way or the other?  As long as we use UUIDs, I don't think either choice will make a significant difference in complexity for the CRUD cycle.  Maybe for filtering, though?

If the user has claims tagged "flying car" and "car accident", and they filter claims by the string "car", then both sets should show up.  I think this pseudocode is sufficiently clean, and it assumes Claim has a list of Tags:

```
for each Claim C
  for each Tag T in C
    show C if T.title matches filterString
```

**Comment by therabidsquirel** [comment]
I agree @kdbanman that Claims should have a list of tags, and this pseudo-code seems sufficiently effective to me.

I'd like to point out a technicality with filtering by tags though, in case I misunderstand. In your example there are two tags, with the strings "flying car" and "car accident", which when filtering by the string "car" should both show up. I thought we were just implementing filtering by whole tags though.

>US03.03.01
As a claimant, I want to filter the list of expense claims **by tags**, to show only those claims that have **at least one tag** matching any of a **given set of one or more filter tags**.

So if "car" was also a tag, neither "flying car" or "car accident" are the tag "car", so filtering by "car" would turn up neither of those. If "car" is not a tag, then you would be unable to filter by it.

**Comment by ryant26** [comment]
@therabidsquirel My interpretation of filtering by tags is the same as yours. 

My opinion of how the data should be structured:

[Types:](http://www.elasticsearch.org/guide/en/elasticsearch/reference/current/glossary.html#glossary-type):
- Claim
- Expense
- Tag
- User

Each Claim, Tag, Expense, and user is it's own [Document.](http://www.elasticsearch.org/guide/en/elasticsearch/reference/current/glossary.html#glossary-document)

The parent-child relationship between: Users-Tags, and Claims-Expenses. (Where Users and Claims are both parents respectively)

I think the best way to associate claims and tags is for the claim to host a list of Tag UUID's.

**Comment by kdbanman** [comment]
Edited.  Thanks for the clarification.

@ryant26, all of that makes sense to me.  Do you agree that the data model objects should inherit from a Document object for cache/remote syncing?

**Comment by ryant26** [comment]
@kdbanman I agree with that, everything I listed as a "Type" should inherit from Document. Are there additional classes I did not list under "Type" that you feel should inherit from Document?

**Comment by kdbanman** [comment]
Nope!  That sounds good to me.  Once that outstanding PR on the data model UML is merged, I can make that change there.  For now, I'll just change the skeleton code to match.

**Comment by TheLastBanana** [comment]
@ryant26's proposal sounds good to me.

**Comment by therabidsquirel** [comment]
@ryant26 also sounds good to me.

**Comment by kdbanman** [comment]
This is what I'll throw in the UML.

```
Document
---------------------
- dirty: boolean
---------------------
- isDirty(): boolean
- setDirty()
- setClean()
```

As far as I can tell, we should be able to hide this from Activity code.  All `setDirty()` calls could come from User, Claim, Item, and Tag setters.  `setClean()` calls can come from the DataSource as it delegates to its saver/loader thing.  Sound feasible?

**Comment by therabidsquirel** [comment]
Looks good.

**Comment by TheLastBanana** [comment]
Yeah, that seems like a good plan.

**Comment by kdbanman** [comment]
Also, all of the things we are concerned about having UUIDs happen to be Documents.  And it really does make sense to uniquely identify documents.  Whether it's filenames or elasticsearch [document ids](http://www.elasticsearch.org/guide/en/elasticsearch/reference/current/mapping-id-field.html).

How does everyone feel about me putting our UUID system right into the Document class?  That doesn't exclude other classes from using UUIDs, but it does mean storage document ids will be the same as the corresponding internal model object ids.

```
Document
---------------------
- docID: UUID
- dirty: boolean
---------------------
+ getUUID(): UUID
+ isDirty(): boolean
+ setDirty()
+ setClean()
```

**Comment by TheLastBanana** [comment]
Sure, that sounds reasonable.

**Comment by therabidsquirel** [comment]
Yep, also sounds good.

**Comment by kdbanman** [comment]
We can refer back to this when it comes time to write Cache / Server Interface classes.  For now, the high level relationships have been implemented in 4cb75d6a00a1660b6dd8cd823e14eb4b891409a3 and described in a969fa8000f425a01403152aae38e2a3a0943256.
